Document FF module configuration (#254)
precice/openfoam-adapter: docs/config.md

The `participant` needs to be the same as the one specified in the `preciceConfig`,
which is the main preCICE configuration file. The `preciceConfig` can be a path and needs to be wrapped with quotation marks.

The list `modules` can contain `CHT`, `FSI`, or/and `FF` (separated by space).

> Each module provides some fields common in the respective type of simulation, and fields from different modules can be combined.

Reviewer: ```suggestion
Each module provides some data fields common in the respective type of simulation, and data fields from different modules can be combined by selecting multiple modules.
```

For example, this is a valid combination: `modules (CHT FF);`.

In the `interfaces`, we specify the coupling interfaces (here only one).
The `mesh` needs to be the same as the one specified in the `preciceConfig`.
The `patches` specifies a list of the names of the OpenFOAM boundary patches that are
participating in the coupled simulation. These need to be defined in the files
included in the `0/` directory. The names of the interfaces (e.g., `Interface1`) are arbitrary and are not used.

> The `locations` field is optional and its default value is `faceCenters` (with `faceCentres` also accepted), signifying that the interface mesh is defined on the cell face centers. The alternative option is `faceNodes`, which defines the mesh on the face nodes and is needed, e.g., for reading displacements in an FSI scenario.

Reviewer: ```suggestion
The `locations` field is optional and its default value is `faceCenters` (with `faceCentres` also accepted), signifying that the interface mesh is defined on the cell face centers. The alternative option is `faceNodes`, which defines the mesh on the face nodes, is needed, e.g., for reading displacements in an FSI scenario.
```

Author: I think this is not what you meant to fix, as it now makes less sense to me.

Reviewer: Ah no, you are right.

The values for `readData` and `writeData`
for conjugate heat transfer
can be `Temperature`, `Heat-Flux`, `Sink-Temperature`,

> or `Heat-Transfer-Coefficient`. Postfixed names like `Sink-Temperature-Domain1` are also allowed.

Reviewer: ```suggestion
or `Heat-Transfer-Coefficient`. Postfixed names like `Sink-Temperature-Domain1` are also allowed (e.g. in order to distinguish multiple data sets of the same type).
```

```c++
readData
(
  Sink-Temperature          // e.g., Sink-Temperature-Solid
  Heat-Transfer-Coefficient // e.g., Heat-Transfer-Coefficient-Solid
);

writeData
(
  Sink-Temperature          // e.g., Sink-Temperature-Fluid
  Heat-Transfer-Coefficient // e.g., Heat-Transfer-Coefficient-Fluid
);
```

> For fluid-structure interaction, `writeData` can be `Force` or `Stress`, where `Stress` is essentially a force vector scaled by the cell face in spatial coordinates (with any postfix), thus, a conservative quantity as well. `readData` can be `Displacement` and `DisplacementDelta` (with any postfix). `DisplacementDelta` refers to the last coupling time step, which needs to considered in the case of subcycling. Structure solvers (such as solidFoam or solids4Foam) can also write `Displacement` and read `Force`.

Reviewer: ```suggestion
For fluid-structure interaction, `writeData` can be `Force` or `Stress`, where `Stress` is essentially a force vector scaled by the cell face in spatial coordinates (with any postfix), thus, a consistent quantity. `readData` can be `Displacement` and `DisplacementDelta` (with any postfix). `DisplacementDelta` refers to the delta of the last coupling time step, which needs to considered in the case of subcycling. Structure solvers (such as solidFoam or solids4Foam) can also write `Displacement` and read `Force`.
```
I'm pretty sure that `Stress` should be mapped consistent. I would also rewrite this a bit: either say initially which fields are supported for Fluid and which for Solid or just list all of them under 'reading' and 'writing'. Right now, it sounds like we had already documentation for a Fluid participant and just tried to add a sentence about the Solid side, which came with new capabilities.

Author: I changes this paragraph to an itemized list, I think this makes it clearer.

#### FSI

* For `readData(Displacement)` or `DisplacementDelta`, you need the following:
  * `type movingWallVelocity` for the interface (e.g., `flap`) in `0/U`,
  * `type fixedValue` for the interface (e.g., `flap`) in the `0/pointDisplacement`, and

>   * `solver displacementLaplacian` in the `constant/dynamicMeshDict`. The solver [`RBFMeshMotionSolver` from solids4foam is also known to work](https://github.com/precice/openfoam-adapter/pull/241), since the OpenFOAM adapter v1.2.0 and the nextRelease branch of solids4foam.

Reviewer: Which one is 'the next one'. This sentence is already doomed to be outdated.

Author: This is how the branch is called, unfortunately:

```suggestion
  * `solver displacementLaplacian` in the `constant/dynamicMeshDict`. The solver [`RBFMeshMotionSolver` from solids4foam is also known to work](https://github.com/precice/openfoam-adapter/pull/241), since the OpenFOAM adapter v1.2.0 and the [`nextRelease` branch (v2) of solids4foam](https://github.com/solids4foam/solids4foam/commit/971d4c08f0171cfd8d53a6f831ce0745d98521de).
```

solver              displacementLaplacian;
```

#### FF

> The fluid-fluid coupling module supports reading and writing pressure, velocity, and their gradients.

Reviewer: It would make sense to spell out the names, as we encoded them in the adapter, as their spelling actually matters for configuration purposes.

Author: Fixed, thanks!

> Similarly to the CHT module, you need a `fixedValue` boundary condition in the respective field in order to read a value,
> and a `fixedGradient` boundary condition in order to read a gradient.

Reviewer: ```suggestion
Similarly to the CHT module, you need a `fixedValue` boundary condition of the respective primary field in order to read and apply values, and a `fixedGradient` boundary condition of the respective gradient field in order to read and apply gradients.
```

{% experimental %}
The FF module is still experimental and the boundary conditions presented here have not been rigorously tested.

> We already have reasons to believe that a `fixedGradient` can have [side-effects](https://github.com/precice/openfoam-adapter/issues/93)
> and may not lead to completely accurate results.

Reviewer: ```suggestion
We already have reasons to believe that a `fixedGradient` can have [side-effects](https://github.com/precice/openfoam-adapter/issues/93) and may not lead to completely accurate results.
```

## Advanced configuration

> These additional parameters may only concern some users is special cases. Keep reading if you want to use nearest-projection mapping, an incompressible or basic (e.g., laplacianFoam) solver, if you are using a solver with different variable names (e.g., a multiphase solver) or if you are trying to debug a simulation.

Reviewer: ```suggestion
These additional parameters may only concern some users in special cases. Keep reading if you want to use the nearest-projection mapping, an incompressible or basic (e.g., laplacianFoam) solver, if you are using a solver with different variable names (e.g., a multiphase solver) or if you are trying to debug a simulation.
```

}
```

> Note that the adapter [does not currently adapt the name of the pressure used in the computations for compressible solvers](https://github.com/precice/openfoam-adapter/issues/253).

Reviewer: ```suggestion
Note that the adapter [does not adapt the name of the pressure used in the computations for compressible solvers](https://github.com/precice/openfoam-adapter/issues/253).
```

Author: But we want to fix that, eventually, no?

Reviewer: Yes, the currently just seemed a bit odd in the sentence.

Use the option `namePointDisplacement unused;` for solvers that do not create a pointDisplacement field, such as the RBFMeshMotionSolver.

For FF simulations:

```c++
FF
{
  // Velocity
  nameU U;
  // Pressure
  nameP p;
}
```

> Note that the adapter does not automatically adapt the pressure name for compressible solvers.

Reviewer: This sentence is a duplicate. On purpose?

Author: Yes, because the one is in the FSI section, and the other in the FF section. But I will see how I can formulate this better.

Author: Clarified, linking to https://www.openfoam.com/documentation/guides/latest/doc/guide-applications-solvers-variable-transform-p-rgh.html
